src/node.ts

```diff
@@ -0,0 +1,651 @@
+        }
+
+        const nodeType = data.slice(offset, offset + nodeForkSizes.nodeType)
+        let nodeForkSize = nodeForkSizes.preReference + refBytesSize
+
+        if (nodeTypeIsWithMetadataType(nodeType[0])) {
+          if (data.length < offset + nodeForkSizes.preReference + refBytesSize + nodeForkSizes.metadata) {
+            throw Error(`Not enough bytes for metadata node fork at byte ${byte}`)
+          }
+
+          const metadataByteSize = fromBigEndian(
+            data.slice(offset + nodeForkSize, offset + nodeForkSize + nodeForkSizes.metadata),
+          )
+          nodeForkSize += nodeForkSizes.metadata + metadataByteSize
+
+          fork = MantarayFork.deserialize(data.slice(offset, offset + nodeForkSize), this.obfuscationKey!, {
+            withMetadata: { refBytesSize, metadataByteSize },
+          })
+        } else {
+          if (data.length < offset + nodeForkSizes.preReference + refBytesSize) {
+            throw Error(`There is not enough size to read fork at offset ${offset}`)
+          }
+
+          fork = MantarayFork.deserialize(data.slice(offset, offset + nodeForkSize), this.obfuscationKey!)
+        }
+
+        this.forks![byte] = fork
+
+        offset += nodeForkSize
+      })
+    } else {
+      throw Error('Wrong mantaray version')
+    }
+  }
+}
+
+export function nodeTypeIsWithMetadataType(nodeType: number): boolean {
+  return (nodeType & NodeType.withMetadata) === NodeType.withMetadata
+}
+
+/**
+ * The hash length has to be 31 instead of 32 that comes from the keccak hash function
+ */
+function serializeVersion(version: MarshalVersion): Bytes<31> {
+  const versionName = 'mantaray'
+  const versionSeparator = ':'
+  const hashBytes = keccak256Hash(versionName + versionSeparator + version)
+
+  return hashBytes.slice(0, 31) as Bytes<31>
+}
+
+function serializeReferenceLength(entry: Reference): Bytes<1> {
+  const referenceLength = entry.length
+
+  if (referenceLength !== 32 && referenceLength !== 64) {
+    throw new Error(`Wrong referenceLength. It can be only 32 or 64. Got: ${referenceLength}`)
+  }
+  const bytes = new Uint8Array(1)
+  bytes[0] = referenceLength
+
+  return bytes as Bytes<1>
+}
+
+/** loads all nodes recursively */
+export async function loadAllNodes(storageLoader: StorageLoader, node: MantarayNode): Promise<void> {
+  if (!node.forks) return
+
+  for (const fork of Object.values(node.forks)) {
+    await fork.node.load(storageLoader, fork.node.getEntry)
+    await loadAllNodes(storageLoader, fork.node)
+  }
+}
+
+/**
+ * Throws an error if the given nodes properties are not equal
+ *
+ * @param a Mantaray node to compare
+ * @param b Mantaray node to compare
+ * @param accumulatedPrefix accumulates the prefix during the recursion
+ * @throws Error if the two nodes properties are not equal recursively
+ */
+export const sameNodes = (a: MantarayNode, b: MantarayNode, accumulatedPrefix = ''): void | never => {
```

Comment:
```suggestion
export const equalNodes = (a: MantarayNode, b: MantarayNode, accumulatedPrefix = ''): void | never => {
```
Similarly to `equalBytes`